OverlayPlugin.Core/NetworkProcessors/OverlayPluginLogLines.cs

```diff
@@ -18,31 +20,129 @@ public OverlayPluginLogLines(TinyIoCContainer container)
 
     class OverlayPluginLogLineConfig
     {
-        private Dictionary<string, Dictionary<string, OpcodeConfigEntry>> config = new Dictionary<string, Dictionary<string, OpcodeConfigEntry>>();
+        private Dictionary<string, Dictionary<string, OpcodeConfigEntry>> opcodes = new Dictionary<string, Dictionary<string, OpcodeConfigEntry>>();
         private ILogger logger;
         private FFXIVRepository repository;
+        private PluginConfig config;
+        private TinyIoCContainer container;
 
         private int exceptionCount = 0;
         private const int maxExceptionsLogged = 3;
 
+        private bool haveAttemptedOpcodeDownload = false;
+
+        private const string remoteOpcodeUrl = "https://raw.githubusercontent.com/OverlayPlugin/OverlayPlugin/main/OverlayPlugin.Core/resources/opcodes.jsonc";
+
         public OverlayPluginLogLineConfig(TinyIoCContainer container)
         {
+            this.container = container;
             logger = container.Resolve<ILogger>();
             repository = container.Resolve<FFXIVRepository>();
-            var main = container.Resolve<PluginMain>();
+            config = container.Resolve<PluginConfig>();
 
-            var pluginDirectory = main.PluginDirectory;
+            // TODO: should we fall back to the file if the remote cached config is somehow broken?
+            if (!LoadCachedOpcodesFromConfig())
+            {
+                LoadOpcodesFromFile();
+            }
+        }
+
+        private bool LoadCachedOpcodesFromConfig()
+        {
+            var versionStr = config.CachedOpcodeOverlayPluginVersion;
+            Version version;
+
+            if (versionStr == "")
+            {
+                return false;
+            }
+
+            try
+            {
+                version = new Version(versionStr);
+            }
+            catch (Exception ex)
+            {
+                LogException($"Invalid CachedOpcodeOverlayPluginVersion {versionStr}: {ex}");
+                ClearCachedOpcodes();
+                return false;
+            }
 
+            // Only load cached opcodes if we have loaded them for this version.
+            // If they're old, clear them.
+            if (repository.GetOverlayPluginVersion() > version)
+            {
+                ClearCachedOpcodes();
+                return false;
+            }
+
+            try
+            {
+                // TODO: is there a better way to go JSON -> Dictionary here?
+                opcodes = JsonConvert.DeserializeAnonymousType(config.CachedOpcodeFile.ToString(), opcodes);

```

Comment:
Aliasing the type might work, I didn't think to test it when I did the original implementation. Something like:

```cs
namespace RainbowMage.OverlayPlugin.NetworkProcessors
{
    using ConfigFileType = Dictionary<string, Dictionary<string, OpcodeConfigEntry>>;
// ...
        private ConfigFileType opcodes = new ConfigFileType();
// ...
                opcodes = JsonConvert.DeserializeObject<ConfigFileType>(jsonData);
```

Reply:
Yeah, mostly I wanted to go JObject/JToken -> Dictionary without going through a string intermediary.  I don't think this makes anything better, as DeserializeObject only takes a string.  I guess I could just save the file as a string, but if we've already parsed it that seems silly to save a JSON string in an otherwise JSON file.  This is probably fine as-is.